cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java

```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
-                                     operationsMap.get(input.getOrigin()), write, visitedOperationNames,
-                                     new HashMap<>());
+  /**
+   * Helper method to compute the outgoing connections
+   * @param currentOperation current operation which needs to evaluated
+   * @param visitedOperations a {@link Set} containing all the operations which has been processed so
+   * far.
+   */
+  private void computeOutgoing(Operation currentOperation, Set<Operation> visitedOperations) {
+    // mark this operation if not already done
+    if (!visitedOperations.add(currentOperation)) {
+      return;
+    }
+
+    // base condition: if the current operation is write we have reached the end
+    if (currentOperation.getType() == OperationType.WRITE) {
+      return;
+    }
+
+    // if this is a transform operation then traverse down to all the outgoing operation from this operation
+    // expanding further the traversal and exploring the operations
+    if (currentOperation.getType() == OperationType.TRANSFORM) {
+      TransformOperation transform = (TransformOperation) currentOperation;
+      Set<Operation> operations = operationOutgoingConnections.get(transform.getName());
+      for (Operation operation : operations) {
+        computeOutgoing(operation, visitedOperations);
       }
     }
+  }
 
-    Set<Operation> fieldOperations = new HashSet<>();
-    for (String operation : visitedOperationNames) {
-      fieldOperations.add(operationsMap.get(operation));
+  /**
+   * Checks whether the given fieldName is used in the next operations or not
+   *
+   * @param previousOperation previous {@link ReadOperation}
+   * @param nextOperation the next operation which should either be a {@link TransformOperation} or {@link
+   * WriteOperation}
+   * @param fieldName the name of the field whose usage needs to be checked
+   * @return true if the field is used in the nextOperation
+   */
+  private boolean isUsed(ReadOperation previousOperation, Operation nextOperation, String fieldName) {
+    if (!previousOperation.getOutputs().contains(fieldName)) {
+      throw new IllegalArgumentException(String.format("The given field '%s' does not exist in the previous operation" +
+                                                         " '%s'", fieldName, previousOperation));
+    }
+    Set<InputField> inputFields = new HashSet<>();
+    if (OperationType.WRITE == nextOperation.getType()) {
+      WriteOperation nextWrite = (WriteOperation) nextOperation;
+      inputFields = new HashSet<>(nextWrite.getInputs());
+    } else if (OperationType.TRANSFORM == nextOperation.getType()) {
+      TransformOperation nextTransform = (TransformOperation) nextOperation;
+      inputFields = new HashSet<>(nextTransform.getInputs());
+    }
+    // if the next operation inputFields does contains the given fieldName return true
+    for (InputField inputField : inputFields) {
```

Comment:
can be simplified to `return inputFields.contains(inputField)`

Reply:
Used: 
```
return inputFields.contains(InputField.of(previousOperation.getName(), fieldName));
```